File: runtime/models/tools.py
"""Pydantic request/response models for the tools API (Phase T0)."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel


class ToolsHealthResponse(BaseModel):
    loaded: bool
    toolCount: int
    categories: dict[str, int] = {}


class ToolCatalogEntry(BaseModel):
    name: str
    title: str
    description: str
    category: str
    permission: str
    latencyClass: str
    enabled: bool
    source: str


class ToolCatalogResponse(BaseModel):
    tools: list[ToolCatalogEntry]


class ToolPlanRequest(BaseModel):
    sessionId: str = ""
    userMessage: str
    agentId: str = "jarvis"
    categories: list[str] | None = None


class ToolPlanResponse(BaseModel):
    useTools: bool
    candidates: list[str] = []
    intent: str = "chat"


class ToolExecuteRequest(BaseModel):
    sessionId: str = ""
    toolName: str
    args: dict[str, Any] = {}
    agentId: str = "jarvis"
    allowedPaths: list[str] | None = None


class ToolExecuteResponse(BaseModel):
    ok: bool
    data: Any | None = None
    error: str | None = None


class ChatTurn(BaseModel):
    role: str
    content: str


class ToolLoopRequest(BaseModel):
    sessionId: str = ""
    userMessage: str
    history: list[ChatTurn] = []
    candidates: list[str] = []
    systemPrompt: str = ""
    agentId: str = "jarvis"
    categories: list[str] | None = None
    allowedPaths: list[str] | None = None
    # Passed through from the browser's AI Settings — never persisted server-side,
    # never written into tool_runs (TOOLS.md §7: no secrets in tool schemas/logs).
    apiKey: str | None = None
    model: str | None = None
    baseUrl: str | None = None
    maxTokens: int = 1024
    temperature: float | None = None


class ToolRunSummary(BaseModel):
    tool: str
    success: bool
    error: str | None = None


class ToolLoopResponse(BaseModel):
    reply: str | None = None
    toolRuns: list[ToolRunSummary] = []
    turns: int = 0
    degraded: bool = False
    reason: str | None = None

File: app/src/services/tools.ts
/**
 * Tools runtime client (Phase T0/T1) — thin fetch wrapper over the Python
 * tool loop, via the Vite `/runtime` proxy. Graceful degradation is the
 * contract: every call catches its own failures and returns a safe default.
 * A dead runtime or a tool error must NEVER throw into the voice flow.
 */

const RUNTIME_BASE = '/runtime'

export interface ToolsHealth {
  loaded: boolean
  toolCount: number
  categories: Record<string, number>
}

export interface ToolCatalogEntry {
  name: string
  title: string
  description: string
  category: string
  permission: string
  latencyClass: string
  enabled: boolean
  source: string
}

export interface ToolPlan {
  useTools: boolean
  candidates: string[]
  intent: string
}

export interface ToolRunSummary {
  tool: string
  success: boolean
  error?: string | null
}

export interface ToolLoopResult {
  reply: string | null
  toolRuns: ToolRunSummary[]
  turns: number
  degraded: boolean
  reason?: string | null
}

export interface ToolExecuteResult {
  ok: boolean
  data?: unknown
  error?: string | null
}

async function post(path: string, body: unknown): Promise<Response | null> {
  try {
    return await fetch(`${RUNTIME_BASE}${path}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    })
  } catch {
    return null
  }
}

export async function getToolsHealth(): Promise<ToolsHealth | null> {
  try {
    const res = await fetch(`${RUNTIME_BASE}/api/tools/health`)
    if (!res.ok) return null
    return (await res.json()) as ToolsHealth
  } catch {
    return null
  }
}

export async function getToolCatalog(
  toolsEnabled: boolean,
  categories?: string[]
): Promise<ToolCatalogEntry[]> {
  try {
    const params = new URLSearchParams({ toolsEnabled: String(toolsEnabled) })
    if (categories?.length) params.set('categories', categories.join(','))
    const res = await fetch(`${RUNTIME_BASE}/api/tools/catalog?${params}`)
    if (!res.ok) return []
    return ((await res.json()) as { tools: ToolCatalogEntry[] }).tools ?? []
  } catch {
    return []
  }
}

const NO_TOOLS: ToolPlan = { useTools: false, candidates: [], intent: 'chat' }

export async function planTools(
  userMessage: string,
  categories?: string[],
  sessionId = ''
): Promise<ToolPlan> {
  const res = await post('/api/tools/plan', { userMessage, sessionId, categories })
  if (!res || !res.ok) return NO_TOOLS
  try {
    return (await res.json()) as ToolPlan
  } catch {
    return NO_TOOLS
  }
}

export interface ToolLoopArgs {
  userMessage: string
  history: { role: string; content: string }[]
  candidates: string[]
  systemPrompt: string
  sessionId?: string
  categories?: string[]
  allowedPaths?: string[]
  apiKey?: string
  model?: string
  baseUrl?: string
  maxTokens?: number
  temperature?: number
}

export async function runToolLoop(args: ToolLoopArgs): Promise<ToolLoopResult | null> {
  const res = await post('/api/tools/loop', args)
  if (!res || !res.ok) return null
  try {
    return (await res.json()) as ToolLoopResult
  } catch {
    return null
  }
}

export async function executeTool(
  toolName: string,
  toolArgs: Record<string, unknown> = {},
  allowedPaths?: string[]
): Promise<ToolExecuteResult> {
  const res = await post('/api/tools/execute', { toolName, args: toolArgs, allowedPaths })
  if (!res || !res.ok) return { ok: false, error: 'runtime unavailable' }
  try {
    return (await res.json()) as ToolExecuteResult
  } catch {
    return { ok: false, error: 'bad response' }
  }
}
